src/app/api/common/delegates/getDelegates.ts

```diff
@@ -13,13 +18,102 @@ import {
   type Delegate,
   type DelegatePayload,
   type DelegatesGetPayload,
+  type DelegateStats,
 } from "./delegate";
 import { fetchIsCitizen } from "../citizens/isCitizen";
 import Tenant from "@/lib/tenant/tenant";
 import { fetchDelegateStatement } from "@/app/api/common/delegateStatement/getDelegateStatement";
 import { fetchCurrentQuorum } from "@/app/api/common/quorum/getQuorum";
 import { fetchVotableSupply } from "@/app/api/common/votableSupply/getVotableSupply";
 
+async function getDelegatesApi(
+  sort: string,
+  pagination: PaginationParamsEx,
+  seed?: number
+): Promise<PaginatedResultEx<any>> {
+  const { namespace } = Tenant.current();
+  const apiDelegatesQuery = (sort: string) =>
+  `
+    SELECT 
+      delegates.delegate,
+      num_of_delegators,
+      direct_vp,
+      avp.advanced_vp,
+      voting_power,
+      contract
+    FROM 
+      ${namespace + ".delegates"}
+    LEFT JOIN
+      ${
+        namespace + ".advanced_voting_power"
+      } avp
+    ON 
+      avp.delegate = delegates.delegate
+    WHERE 
+      num_of_delegators IS NOT NULl
+    ORDER BY 
+      ${sort}
+    OFFSET $1
+    LIMIT $2;
+  `;
+  // TODO: voting power sort, others
+  const paginatedQuery = async (skip: number, take: number) => {
+    switch (sort) {
+      case "most_delegators":
+        return prisma.$queryRawUnsafe<DelegatesGetPayload[]>(
+          apiDelegatesQuery("num_of_delegators DESC"),
+          skip,
+          take
+        );
+      case "weighted_random":
+        await prisma.$executeRawUnsafe(`SELECT setseed($1);`, seed);
+        return prisma.$queryRawUnsafe<DelegatesGetPayload[]>(
+          apiDelegatesQuery("-log(random()) / voting_power"),
+          skip,
+          take
+        );
+      default:
+        throw new Error("Invalid sort order");
+    }
+  };
+
+  const { meta, data: delegates } = await paginateResultEx(
+    paginatedQuery,
+    pagination
+  );
+  const _delegates = await Promise.all(
+    delegates.map(async (delegate: DelegatesGetPayload) => {
+      return {
+        citizen: await fetchIsCitizen(delegate.delegate),
```

Comment:
I think `isCitizen` check one can be incorporated into the main `apiDelegatesQuery` since the data is coming from the same DB

Reply:
Also, thinking for the future here, delegate statements are going to live in the same DB as well. So, ideally, we can join all this data in the same query

Reply:
done, touched that up